pallets/dapp-staking-v3/src/types.rs

```diff
@@ -638,4 +810,463 @@ impl EraInfo {
+        match idx {
+            Ok(idx) => self.0.get(idx).map(|x| *x),
+            Err(idx) => {
+                if idx.is_zero() {
+                    None
+                } else {
+                    match self.0.get(idx - 1) {
+                        Some(info) if info.period() == period => {
+                            let mut info = *info;
+                            info.era = era;
+                            Some(info)
+                        }
+                        _ => None,
+                    }
+                }
+            }
+        }
+    }
+
+    /// Last era for which a stake entry exists, `None` if no entries exist.
+    pub fn last_stake_era(&self) -> Option<EraNumber> {
+        if let Some(last_element) = self.0.last() {
+            Some(last_element.era())
+        } else {
+            None
+        }
+    }
+
+    /// Last period for which a stake entry exists, `None` if no entries exist.
+    pub fn last_stake_period(&self) -> Option<PeriodNumber> {
+        if let Some(last_element) = self.0.last() {
+            Some(last_element.period())
+        } else {
+            None
+        }
+    }
+
+    pub fn total_staked_amount(&self, period: PeriodNumber) -> Balance {
+        match self.0.last() {
+            Some(last_element) if last_element.period() == period => {
+                last_element.total_staked_amount()
+            }
+            _ => Balance::zero(),
+        }
+    }
+
+    pub fn staked_amount(&self, period: PeriodNumber, period_type: PeriodType) -> Balance {
+        match self.0.last() {
+            Some(last_element) if last_element.period() == period => {
+                last_element.staked_amount(period_type)
+            }
+            _ => Balance::zero(),
+        }
+    }
+
+    /// Stake the specified `amount` on the contract, for the specified `period_type` and `era`.
+    pub fn stake(
+        &mut self,
+        amount: Balance,
+        period_info: PeriodInfo,
+        era: EraNumber,
+    ) -> Result<(), ()> {
+        // Defensive check to ensure we don't end up in a corrupted state. Should never happen.
+        if let Some(last_element) = self.0.last() {
+            if last_element.era() > era || last_element.period() > period_info.number {
+                return Err(());
+            }
+        }
+
+        // Get the most relevant `ContractStakingInfo` instance
+        let (last_element_has_matching_era, last_element_has_matching_period) =
+            if let Some(last_element) = self.0.last() {
+                (
+                    last_element.era() == era,
+                    last_element.period() == period_info.number,
+                )
+            } else {
+                (false, false)
+            };
+
+        // Prepare the new entry
+        let mut staking_info = if last_element_has_matching_era {
+            self.0.remove(self.0.len() - 1)
```

Comment:
```suggestion
            self.0.pop()
```

Reply:
IMO it looks clean and don't panic (there is more than one instance of this, L1222, etc)

Reply:
I see, then it's fine the way it is.